src/core/regimes.rkt

```diff
@@ -175,127 +175,131 @@
+       ;; the current alt-error-sum is less then previous
+       (when (or (not (vector-ref best-alt-idx prev-split-idx))
+                 (fl< current-error (vector-ref best-alt-cost prev-split-idx)))
+        ;; update best cost and best index
+        (vector-set! best-alt-cost prev-split-idx current-error)
+        (vector-set! best-alt-idx prev-split-idx alt-idx))))))
 
-  ;; prefix of p is for previous
-  ;; prefix of n is for next
-  ;; prefix of f is for final result vectors
-  ;; a for acost vectors
-  ;; b for candidate index
-  ;; c for alt index
-  ;; d for previous index
-  ;; This is where the high level bulk of the algorithm is applied
-  ;; We get the final splitpoints by applying add-splitpoints as many times as we want
-  (define-values (pa pb pd) (initial))
-  (define-values (fa fb fd)
-    ; short circuit if there is no other alts to consider
-    (if (> num-candidates 1)
-      (let loop ([pa pa] [pb pb] [pd pd])
-      (define-values (na nb nd) (add-splitpoint pa pb pd))
-      (if (equal? nb pb) ;; only need to compare candidate index
-          (values na nb nd)
-          (loop na nb nd)))
-    (values pa pb pd)))
-    
-    ;; From here down is messy code translating from 4 vectors back to
-    ;; the original list of split points
-    (define fixed-final (make-vector num-points))
-    (define fp (list->vector (range 1 (+ num-points 1))))
-    (for ([idx (in-range 0 num-points)])
-      (define a (flvector-ref fa idx))
-      (define b (vector-ref fb idx))
-      (define c (vector-ref fp idx))
-      (define d (vector-ref fd idx))
-      (vector-set! fixed-final idx (cand a b c d)))
+   ;; Save current values for the current point we are working on.
+   (define current-alt-error (flvector-ref result-error-sums point-idx))
+   (define current-alt-idx (vector-ref result-alt-idxs point-idx))
+   (define current-prev-idx (vector-ref result-prev-idxs point-idx))
+   ;; We have now have the index of the best alt and it's error up to our 
+   ;; current point-idx.
+   ;; Now we compare against our current best saved in the 3 vectors above
+   (for ([prev-split-idx (in-range 0 point-idx)])
+    (when (vector-ref can-split-vec (+ prev-split-idx 1))
+     ;; Re compute the error sum for a potential better alt
+     (define alt-error-sum (fl+ (fl+ (flvector-ref result-error-sums prev-split-idx)
+                    (vector-ref best-alt-cost prev-split-idx)) min-weight))
+     ;; pre-compute values for tie breaking
+     (define adjusted-cost current-alt-error)
+     (define current-best-alt-idx (vector-ref best-alt-idx prev-split-idx))
+     ;; Check if the new alt-error-sum is better then the current
+     (define set-cond
+      ;; give benefit to previous best alt
+      (cond [(fl< alt-error-sum adjusted-cost) #t]
+            ;; Tie breaker if error are the same favor first alt
+            [(and (fl= alt-error-sum adjusted-cost)
+                  (> current-alt-idx current-best-alt-idx)) #t]
```

Comment:
The usual way you do this is something like:
```
(cond
 [(< a b) #t]
 [(> a b) #f]
 [(< c d) #t]
 [(> c d) #f]
  ...)
```
This avoids testing the `=` branch over and over again.

Reply:
Ya I thought it was pretty ugly but was like well this "works". Moving on. Will fix though

Reply:
still working on this